clippy_lints/src/dereference.rs

```diff
@@ -920,6 +941,150 @@ fn ty_contains_infer(ty: &hir::Ty<'_>) -> bool {
+fn needless_borrow_impl_arg_position<'tcx>(
+    cx: &LateContext<'tcx>,
+    parent: &Expr<'tcx>,
+    arg_index: usize,
+    param_ty: ParamTy,
+    mut expr: &Expr<'tcx>,
+    precedence: i8,
+) -> Position {
+    let sized_trait_def_id = cx.tcx.lang_items().sized_trait();
+
+    let Some(callee_def_id) = fn_def_id(cx, parent) else { return Position::Other(precedence) };
+    let substs_with_expr_ty = cx
+        .typeck_results()
+        .node_substs(if let ExprKind::Call(callee, _) = parent.kind {
+            callee.hir_id
+        } else {
+            parent.hir_id
+        });
+
+    let predicates = cx.tcx.param_env(callee_def_id).caller_bounds();
+    let trait_predicates = predicates
+        .iter()
+        .filter_map(|predicate| {
+            if let PredicateKind::Trait(trait_predicate) = predicate.kind().skip_binder()
+                && is_substs_for_type(trait_predicate.trait_ref.substs, param_ty.to_ty(cx.tcx))
+            {
+                Some(trait_predicate)
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<_>>();
+    let projection_predicates = predicates
+        .iter()
+        .filter_map(|predicate| {
+            if let PredicateKind::Projection(projection_predicate) = predicate.kind().skip_binder()
+                && is_substs_for_type(projection_predicate.projection_ty.substs, param_ty.to_ty(cx.tcx))
+            {
+                Some(projection_predicate)
+            } else {
+                None
+            }
+        })
+        .collect::<Vec<_>>();
+
+    // If no traits were found, or only the `Sized` or `Any` traits were found, return.
+    if trait_predicates.iter().all(|trait_predicate| {
+        let trait_def_id = trait_predicate.def_id();
+        Some(trait_def_id) == sized_trait_def_id || cx.tcx.is_diagnostic_item(sym::Any, trait_def_id)
+    }) {
+        return Position::Other(precedence);
+    }
+
+    // If `param_ty` appears in more than one place, then substituting it with `T` instead of `&T` could
+    // cause a type error.
+    if cx
+        .tcx
+        .fn_sig(callee_def_id)
+        .skip_binder()
+        .inputs_and_output
+        .iter()
+        .enumerate()
+        .any(|(i, ty)| i != arg_index && contains_ty(ty, param_ty.to_ty(cx.tcx)))
+    {
+        return Position::Other(precedence);
+    }
+
+    let check_referent = |referent| {
+        let referent_ty = cx.typeck_results().expr_ty(referent);
+
+        if !is_copy(cx, referent_ty) {
+            return false;
+        }
+
+        // https://github.com/rust-lang/rust-clippy/pull/9136#pullrequestreview-1037379321
+        if !matches!(referent_ty.kind(), ty::Ref(_, _, Mutability::Mut))
+            && trait_predicates
+                .iter()
+                .any(|TraitPredicate { trait_ref, .. }| has_ref_mut_self_method(cx, trait_ref.def_id))
+        {
+            return false;
+        }
+
+        let mut substs_with_referent_ty = substs_with_expr_ty.to_vec();
```

Comment:
The `Vec` can be constructed outside the closure. The same item is being modified in every call.